extensions/wikia/EmbeddableDiscussions/EmbeddableDiscussionsController.class.php

```diff
@@ -36,134 +33,170 @@ public static function onBeforePageDisplay() {
+					wfMessage( 'embeddable-discussions-parameter-error-range', self::ITEMS_MIN, self::ITEMS_MAX )->plain()
+				)->plain();
 
 				return false;
 			}
 		}
 
 		// columns must be integer in range
-		if ( isset( $args['columns'] ) ) {
-			$columns = $args['columns'];
+		if ( isset( $args[ 'columns' ] ) ) {
+			$columnsArg = $args[ 'columns' ];
+			$columns = ctype_digit( $columnsArg ) ? intval( $columnsArg ) : $columnsArg;
 
-			if ( !ctype_digit( $columns ) ||
-				intval( $columns ) > self::COLUMNS_MAX ||
-				intval( $columns ) < self::COLUMNS_MIN
+			if ( !is_int( $columns ) ||
+				$columns > self::COLUMNS_MAX ||
+				$columns < self::COLUMNS_MIN
 			) {
-				// TODO: Refactor error messages to avoid direct concatenation, see JPN-657
-				$errorMessage = wfMessage( 'embeddable-discussions-parameter-error', 'columns' )->plain() .
-					wfMessage( 'embeddable-discussions-parameter-error-range', self::COLUMNS_MIN , self::COLUMNS_MAX )->plain();
+				$errorMessage = wfMessage( 'embeddable-discussions-parameter-error',
+					'columns',
+					wfMessage( 'embeddable-discussions-parameter-error-range', self::COLUMNS_MIN, self::COLUMNS_MAX )->plain()
+				)->plain();
 
 				return false;
 			}
 		}
 
 		// category must be a valid category
-		if ( $modelData['invalidCategory'] ) {
-			// TODO: Refactor error messages to avoid direct concatenation, see JPN-657
-			$errorMessage = wfMessage( 'embeddable-discussions-parameter-error', $args['category'] )->plain() .
-				wfMessage( 'embeddable-discussions-parameter-error-category' )->plain();
+		if ( !empty( $modelData[ 'invalidCategory' ] ) ) {
+			$errorMessage = wfMessage( 'embeddable-discussions-parameter-error', $args[ 'category' ],
+				wfMessage( 'embeddable-discussions-parameter-error-category' )->plain()
+			)->plain();
 
 			return false;
 		}
 
 		return true;
 	}
 
+	/**
+	 * @param array $args
+	 * @return string
+	 */
+	public function render( array $args ) {
+		global $wgCityId;
+
+		$showLatest = !empty( $args[ 'mostrecent' ] ) && filter_var( $args[ 'mostrecent' ], FILTER_VALIDATE_BOOLEAN );
+		$itemCount = !empty( $args[ 'size' ] ) ? intval( $args[ 'size' ] ) : self::ITEMS_DEFAULT;
+		$columns = !empty( $args[ 'columns' ] ) ? intval( $args[ 'columns' ] ) : self::COLUMNS_DEFAULT;
+		$category = !empty( $args[ 'category' ] ) ? $args[ 'category' ] : '';
+
+		$modelData = ( new DiscussionsThreadModel( $wgCityId ) )->getData( $showLatest, $itemCount, $category );
+
+		if ( !$this->checkArguments( $modelData, $args, $errorMessage ) ) {
+			return $this->renderError( $errorMessage );
+		}
+
+		return F::app()->checkSkin( 'wikiamobile' ) ?
+			$this->renderMobile( $modelData, $showLatest, $itemCount ) :
+			$this->renderDesktop( $modelData, $showLatest, $category, $columns );
+	}
+
+	/**
+	 * @param $errorMessage
+	 * @return string
+	 */
+	private function renderError( $errorMessage ) {
+		$templateEngine = ( new Wikia\Template\MustacheEngine )->setPrefix( __DIR__ . '/templates' );
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "/templates" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1192)